http/client: overrides fmt formatter for request to hide sensitive info

abhijat opened the pull request:
## Cover letter

An http request can contain headers which have sensitive information such as authorization. We redact these fields right now using a special function which creates a copy of the request with sensitive fields removed. This has the downside of making the developer remember to call the function each time the request is printed.

The changes here override the fmtlib formatter for request so that any objects logged via fmt are automatically redacted. It is not possible to override the ostream operator without subclassing the request class and using that subclass everywhere, using the fmt API is less intrusive to the code base.

fixes #5011

## Backport Required

- [ ] not a bug fix
- [x] papercut/not impactful enough to backport
- [ ] v22.2.x
- [ ] v22.1.x
- [ ] v21.11.x

A backport is not explicitly required because the current group of sensitive headers is already handled in the code base. 

## UX changes

None

## Release notes

None

abhijat commented:
> Nice. Would it be possible to add a simple test that checks the logs and ensures the redaction works. I think all the required functionality is already in the test tree.

I adjusted an existing test that I had written earlier to exercise this code here: https://github.com/redpanda-data/redpanda/blob/9e4388fe5195f1dda67d2edf2603931dc42a7ae0/src/v/http/tests/http_client_test.cc#L858 

It asserts that redaction works with `fmt::format`. Does this seem sufficient or should I add an explicit test for vlog?

VladLazar commented:
> > Nice. Would it be possible to add a simple test that checks the logs and ensures the redaction works. I think all the required functionality is already in the test tree.
> 
> I adjusted an existing test that I had written earlier to exercise this code here:
> 
> https://github.com/redpanda-data/redpanda/blob/9e4388fe5195f1dda67d2edf2603931dc42a7ae0/src/v/http/tests/http_client_test.cc#L858
> 
> It asserts that redaction works with `fmt::format`. Does this seem sufficient or should I add an explicit test for vlog?

Sorry I missed that. I was looking for a ducktape test and my eyes glazed over the unit test change. I think this is sufficient as it'spretty hard to misuse.

abhijat commented:
> I adjusted an existing test that I had written earlier to exercise this code here:
> 
> https://github.com/redpanda-data/redpanda/blob/9e4388fe5195f1dda67d2edf2603931dc42a7ae0/src/v/http/tests/http_client_test.cc#L858
> 
> It asserts that redaction works with `fmt::format`. Does this seem sufficient or should I add an explicit test for vlog?

On further thought I have adjusted the test to use the seastar logger and not use fmtlib directly, if it's underlying implementation changes to use something other than `fmt` we can be warned by the test.

dotnwat commented on src/v/http/client.h:
```diff
@@ -258,3 +258,20 @@ inline ss::future<> client::forward(client* client, BufferSeq&& seq) {
 } // namespace http
+
+template<>
+struct fmt::formatter<http::client::request_header> {
+    constexpr auto parse(fmt::format_parse_context& ctx)
+      -> decltype(ctx.begin()) {
+        return ctx.begin();
+    }
+
+    template<typename FormatContext>
+    auto format(const http::client::request_header& h, FormatContext& ctx)
+      -> decltype(ctx.out()) {
+        auto redacted = http::redacted_header(h);
+        std::stringstream s;
+        s << redacted;
+        return fmt::format_to(ctx.out(), "{}", s.str());
+    }
```
neat
